applications/asset_tracker_v2/CMakeLists.txt

```diff
@@ -39,3 +39,8 @@ if (CONFIG_BOARD_QEMU_X86 OR CONFIG_BOARD_NATIVE_POSIX)
         target_compile_options(app PRIVATE
 	        -DCONFIG_LTE_NEIGHBOR_CELLS_MAX=10)
 endif()
+
+# Enable Link Time Optimization for the application code
+if(CMAKE_C_COMPILER_ID STREQUAL "GNU" AND CONFIG_SIZE_OPTIMIZATIONS)
+        target_compile_options(app PRIVATE "-flto")
```

Comment:
applying this locally on an app is a good starting point.

Using it solely on sources in app shall still link other Zephyr libraries as regular libs.
Just note that `lto` on linking will be inferred:
> GCC automatically performs link-time optimization if any of the objects involved were compiled with the -flto command-line option. 

Remember to careful investigated that referenced but KEEP sections in the object file are not pruned, for example functions referenced through SYS_INIT.

Reply:
I have investigated this in two ways.

- I had a look at the object dump of zephyr.elf. I noticed that the functions referenced through SYS_INIT do get included in the final binary. The structs named `__init_FUNCTION` do get an entry in the final hex.  The `__used` keyword in https://github.com/zephyrproject-rtos/zephyr/blob/44e691e20e0b323ad67227ee072bd4166c75d894/include/zephyr/init.h#L189 ensures this.
- Performed a test locally to see that the button presses do get sent to the cloud. The button initialization happens during init (they are registered using SYS_INIT here https://github.com/nrfconnect/sdk-nrf/blob/a46fec90210e725382d8eded1871013e0561ac1a/applications/asset_tracker_v2/src/modules/ui_module.c#L659 )